### ISSUE TITLE
High variation in mIOU for different evaluations of same model

### ISSUE THREAD

**Issue by mustafa1728** [MendelXu/zsseg.baseline#13]
Hello authors, thank you for this great work!

I was trying to run for PASCAL VOC following instructions in the readme. I used the single `_single_prompt_` version of config and checked for both `CLIP_ADAPTER` and `REGION_CLIP_ADAPTER` these parameters:
~~~yaml
PROMPT_LEARNER: predefined
PREDEFINED_PROMPT_TEMPLATES:
    - a sculpture of a {}.
~~~

After a single training, for different runs of eval (using the `--eval-only` and `--resume` args) with different seeds, I am getting highly varying numbers. For some two runs, I saw a difference of around 20 points for mIOU-unbase. The trained model is not changing so this seemed very strange.

Can you please help me understand if Im making some mistake or if this is expected behaviour, what is the source for this high variation in evaluation? Initial thoughts were that randomly selected prompts were the source, but later checked that it was actually using the single prompt.

**Comment by prinshul** [comment]
Hi @MendelXu 
I am facing a similar issue. If we run the evaluation multiple times with different seeds, the performance varies on each evaluation with Single Prompt (Fixed Prompt) even though the trained model is the same. Can you please explain the reason for this randomness in results? Thank you.

**Comment by MendelXu** [comment]
Thanks for your report. I didn't notice that before. And I will have a try.

**Comment by prinshul** [comment]
> Thanks for your report. I didn't notice that before. And I will have a try.

Sure. Please let me about your observations as well. Thanks.

**Comment by MendelXu** [comment]
As my experiments results are stored on the internal cloud service. So I test method 1 firstly.
- 1 : Load weights directly.
  Command: 
```
for i in 1 2 3; do python train_net.py --eval-only  --config-file configs/voc-11k-15/zero_shot_maskformer_R101c_single_prompt_bs16_20k.yaml --num-gpus 8 MODEL.WEIGHTS ./output/model_2022_7_25_down.pth > output/voc_log_${i}.txt;done
```
  Seed:  43817718,50551367,39714311
 Results: 64.9615525922651,64.9615525922651,64.9615525922651

It seems normal. I will submmit some jobs to test like yours.

**Comment by mustafa1728** [comment]
Thank you @MendelXu for your reply.

Please find the trained model and the config I used here: https://drive.google.com/drive/folders/17J1nJ07PoyECXBnCJsE5LNPb2uCMUy_D?usp=sharing.

The command I used was 
~~~
python3 train_net.py --config-file config.yaml --eval-only --resume
~~~
The model weight was saved as `model_final.pth` in `output`, the output directory for the config.

**Comment by MendelXu** [comment]
It is wierd. I can always get 75.5977113208025 though I have run it for three times.

**Comment by MendelXu** [comment]
Could you post your log here? I have uploaded my evaluation log here. https://drive.google.com/drive/folders/1pCIVJTklqBTFmFv4fEvNNGSyKgdnAsbh?usp=sharing
You may check if there are some difference.

**Comment by mustafa1728** [comment]
Here are logs for three runs: https://drive.google.com/drive/folders/1llCTMHJ6Nz4A99z3bNl7-97Zf8GsK-t_?usp=sharing. Thank you.

**Comment by mustafa1728** [comment]
In your log's printed configs, there seem to be some extra parameters from mine.

~~~yaml
CLIP_NET:
    FROZEN_NORM: true
    NORM: BN
~~~

I think these are not present anywhere in the repo. Could you please check this and update? Is the variation because of unfrozen norm layers? 

Thank You.

**Comment by MendelXu** [comment]
This part shouldn't affect the performance as it only correspond to the mask former with clip pretrained backbone. 
I think the problem is here. ( snapshot from your  provided log.)
<img width="1227" alt="1658782892607" src="https://user-images.githubusercontent.com/44802022/180873502-4b4b5023-a44a-40b3-a79d-73e0db5f9ef0.png">
**For some unknown reason, the model didn't load some parameters and initialize them randomly.** However, I have tested your provided model with the same code( I clone it from here directly), the result is still normal. Maybe you can load the checkpoint manually and check whether there exists "pixel_decoder.pixel_decoder.*" in it.

**Comment by mustafa1728** [comment]
I used this code to print relevant parameter names in the checkpoint, 

~~~python
import torch

weights = torch.load("model_final.pth")

weight_names = weights["model"].keys()
weight_names_filtered = [n for n in weight_names if "pixel_decoder" in n]
for weight_name in weight_names_filtered:
    print(weight_name)

~~~

The log from running this is
~~~
sem_seg_head.pixel_decoder.adapter_1.weight
sem_seg_head.pixel_decoder.adapter_1.norm.weight
sem_seg_head.pixel_decoder.adapter_1.norm.bias
sem_seg_head.pixel_decoder.layer_1.weight
sem_seg_head.pixel_decoder.layer_1.norm.weight
sem_seg_head.pixel_decoder.layer_1.norm.bias
sem_seg_head.pixel_decoder.adapter_2.weight
sem_seg_head.pixel_decoder.adapter_2.norm.weight
sem_seg_head.pixel_decoder.adapter_2.norm.bias
sem_seg_head.pixel_decoder.layer_2.weight
sem_seg_head.pixel_decoder.layer_2.norm.weight
sem_seg_head.pixel_decoder.layer_2.norm.bias
sem_seg_head.pixel_decoder.adapter_3.weight
sem_seg_head.pixel_decoder.adapter_3.norm.weight
sem_seg_head.pixel_decoder.adapter_3.norm.bias
sem_seg_head.pixel_decoder.layer_3.weight
sem_seg_head.pixel_decoder.layer_3.norm.weight
sem_seg_head.pixel_decoder.layer_3.norm.bias
sem_seg_head.pixel_decoder.layer_4.weight
sem_seg_head.pixel_decoder.layer_4.norm.weight
sem_seg_head.pixel_decoder.layer_4.norm.bias
sem_seg_head.pixel_decoder.mask_features.weight
sem_seg_head.pixel_decoder.mask_features.bias
~~~

So it seems the weights are normal and no `pixel_decoder.pixel_decoder` is present in them. Ill check the part in detectron2 where it loads the model weights.

**Comment by mustafa1728** [comment]
I believe the problem was because of these lines:
https://github.com/MendelXu/zsseg.baseline/blob/bb6504aa8b573de1afa7f58c8e5d40de524f0a58/mask_former/modeling/heads/zero_shot_mask_former_head.py#L40-L41

I changed it to:
~~~python
if "sem_seg_head" in k and not k.startswith(prefix + "predictor") and not "pixel_decoder" in k:
    newk = k.replace(prefix, prefix + "pixel_decoder.")
~~~


Now the results for different runs are consistent.

Thanks a lot for helping to understand and resolve the issue @MendelXu!

**Comment by AnukritiSinghh** [comment]
Hey @mustafa1728, when running proposal classification training for learned prompt for PASCAL VOC dataset, did you face an error like this? 
<img width="1098" alt="image" src="https://user-images.githubusercontent.com/28527591/181755982-309ea29f-8286-4c2d-bbd9-0c460d188717.png">

**Comment by mustafa1728** [comment]
Hi @AnukritiSinghh.

Yes, I believe this error pops up because the classification dataset `voc_base_sem_seg_train_classification` actually has all classes present in it.

It got fixed by filtering out only base classes in the preprocessing script here: https://github.com/MendelXu/zsseg.baseline/blob/bb6504aa8b573de1afa7f58c8e5d40de524f0a58/tools/mask_cls_collect.py#L13-L16
